### PR TITLE
Added an option for links to not allow linking to entries

### PR DESCRIPTION
**Pull request by DrunkHronus**
Added an option 'linkToEntry' to toggle if linking allows selecting entries.  
By default it is true and doesn't change anything, but if set to false, it will not show the option to select entries when linking.

### CODE REVIEW

**Comment by joltmode** [comment]
Ha, great!

Was about to implement this myself.

I am using Craft as a fully Headless, API first CMS with React frontend and the linking to entry feature, due to leaving a static link is causing more trouble than good.

The UX for content editors is a top priority here and removing the affordance to link in content is a godsent!

Though, seeing how you have implemented this, I was thinking that it would be more future-proof and less conflicting if the setting would be moved in a `craftcms` or other namespace within the config?

**Comment by andris-sevcenko** [comment]
> I am using Craft as a fully Headless, API first CMS with React frontend and the linking to entry feature, due to leaving a static link is causing more trouble than good.

Not sure what you mean by that, as inserting an entry via Redactor saves an element reference to the database (in the form of `{entry:ID:url}`, which, in turn, gets turned into a link when fetched from the database.

As for the pull request, we'll discuss this internally, but I feel like that should be a per-field setting, instead of being a per-config setting. It's entirely possible that you'd want to reuse the Redactor config across multiple fields, while choosing not to enable entry linking for some of them.

**Comment by brandonkelly** [comment]
I’m not sure I understand the point of this. Redactor will only show entries that are both live and have URLs. So even with this change implemented, it’s still possible to link to your entries – just by typing their URLs in manually.

So if you don’t want your entries to be linkable, you should disable their URLs from the section’s settings.

And if you want to prevent authors from being able to create hyperlinks at all within your Redactor field, just remove `"link"` from your [Redactor config’s](https://github.com/craftcms/redactor#redactor-configs) `buttons` array.

**Comment by joltmode** [comment]
Per-field setting actually makes sense.

Elaboration for previous comment
----------------------------------

***TL;DR:***

- CraftQL causes some problems to use "Link to Entry" properly, will PR them.
- "Link to Entry" does not give an option to limit what types of entries (and types of entry types) one can link to; currently it's best to prevent "Linking to Entries" in general (**this PR**) to not cause any hiccups for content editors. (Entries Subset plugin is a great place to look for when referencing with limits)
  - Required when project has one entry type for "pages" content.
- In fully component based React world, it makes more sense to not use entry links, but reference entries directly and provide even richer experience for visitors.
  - If links are required, one can still copy-paste the absolute url.

***Long version:***

As for the unusable links, I got put away initially when I found out how it sets up the absolute link with `#<ref>` and thought that it would not make sense to me as I would have to parse it.

After my comment yesterday, I went on to learn how redactor works with Craft and found out how refs work and that they are usable.

Though, then there is another limitation, rooted in CraftQL plugin... 
See, in my stack, Craft is more like a content microservice. I have set it up with CraftQL, and I just query it for data. All of the view logic is separated into a React application that lives even on a different host.

Meaning, that whenever I query for the data, the link being returned is an absolute URL to the entry, which isn't applicable in my frontend logic, as I have a custom routing structure and custom base URL.
Then I remembered that I can set up a custom URL for Site in Craft, but apparently, CraftQL, when loading it's GraphiQL interface, relies on the Site URL and if it's pointed to the other host, it fails.
No biggie, the API itself should still work, but I use the GraphiQL to test queries.

If I was to use this mechanism as it is, then, upon receiving the data, I have to parse it for links and then replace the links to my structure... and working with raw HTML is highly discouraged in React.

---

Then, there is the other problem, where not all of the entry types are meant to be viewable entries per se. That's a different issue, though, but I would love a way to limit what entries I can link to, including what types of the entry, as in https://github.com/nfourtythree/craft3-entriessubset,

Some of the entries in my setup just act as a distinct data containers which to reference in other entries. I have a structure type section "Pages" where I use "content" Matrix field with different block presets, including blocks for entry references, to then be used as data for components in frontend.

---

While I will go on to fix the CraftQL issue with external Site URL, the "Link to Entry" feature as it is now, does not cut it anyways due to unlimited entry selection. Which backfeeds in a positive way for this solution, to remove the entry linking option from redactor type fields.

---

In my setup, the content is structured with Matrix blocks, as with blocks I can refer to them as preset components in React world.
Rich content (WYSWIYG) is also split up in blocks: paragraph, image, title, subtitle ... because those also are specific reusable components in React.
There, the paragraph block type, uses `breakline: true` option, to return unwrapped (no `<p>`, no `<div>`) content and is limited to "safe" inline styles only like bolding, italics, linking (and if need be to link to actual page, the content editor can simply copy the link to the content), etc. Then, I have a distinct RichParagraph component, that is taught to use raw HTML, which I know to be safe, to be put as it is in React.
This gives me the power, to not only add link to entry in WYSIWYG area, but I can get somewhat a rich, "widget-y" entry representation.
